nautilus/include/nautilus/val.hpp

```diff
@@ -348,17 +359,21 @@ namespace details {
 #define DEFINE_BINARY_OPERATOR_HELPER(OP, OP_NAME, OP_TRACE, RES_TYPE)                                                                                                                                                                         \
 	template <typename LHS, typename RHS>                                                                                                                                                                                                      \
 	auto inline OP_NAME(LHS&& left, RHS&& right) {                                                                                                                                                                                             \
-		typedef typename std::common_type<typename LHS::basic_type, typename RHS::basic_type>::type commonType;                                                                                                                                \
-		auto&& lValue = cast_value<LHS, commonType>(std::forward<LHS>(left));                                                                                                                                                                  \
-		auto&& rValue = cast_value<RHS, commonType>(std::forward<RHS>(right));                                                                                                                                                                 \
-		using resultType = decltype(getRawValue(lValue) OP getRawValue(rValue));                                                                                                                                                                  \
-		if SHOULD_TRACE () {                                                                                                                                                                                                                   \
-			auto tc = tracing::traceBinaryOp<tracing::OP_TRACE, resultType>(details::getState(lValue), details::getState(rValue));                                                                                                             \
-			return RES_TYPE(tc);                                                                                                                                                                                                               \
+		if constexpr (!requires(LHS l, RHS r) { getRawValue(l) OP getRawValue(r); }) {                                                                                                                                                                                   \
```

Comment:
Why do we need this and the runtime exception?

Reply:
We need this due to using the visitor pattern and as not all operations are defined on all possible combinations of data types.

Reply:
have you reverted this change?